pkg/admission/kubequota/kubequota_admission.go

```diff
@@ -0,0 +1,359 @@
+}
+
+// TODO(ncdc): find a way to support the special evaluators for pods/pvcs/services.
+// var (
+// 	podsGVR                   = corev1.SchemeGroupVersion.WithResource("pods")
+// 	servicesGVR               = corev1.SchemeGroupVersion.WithResource("services")
+// 	persistentVolumeClaimsGVR = corev1.SchemeGroupVersion.WithResource("persistentvolumeclaims")
+// )
+//
+// type hackyAdmissionAttributesForBuiltInTypes struct {
+// 	admission.Attributes
+// 	builtInCurrent runtime.Object
+// 	builtInOld     runtime.Object
+// }
+//
+// func (h *hackyAdmissionAttributesForBuiltInTypes) GetObject() runtime.Object {
+// 	if h.builtInCurrent != nil {
+// 		return h.builtInCurrent
+// 	}
+//
+// 	obj, converted := objectFor(h.GetResource(), h.Attributes.GetObject())
+// 	if converted {
+// 		h.builtInCurrent = obj
+// 	}
+//
+// 	return obj
+// }
+//
+// func (h *hackyAdmissionAttributesForBuiltInTypes) GetOldObject() runtime.Object {
+// 	if h.builtInOld != nil {
+// 		return h.builtInOld
+// 	}
+//
+// 	obj, converted := objectFor(h.GetResource(), h.Attributes.GetOldObject())
+// 	if converted {
+// 		h.builtInOld = obj
+// 	}
+//
+// 	return obj
+// }
+//
+// func objectFor(gvr schema.GroupVersionResource, obj runtime.Object) (out runtime.Object, converted bool) {
+// 	switch gvr {
+// 	case podsGVR:
+// 		pod := &corev1.Pod{}
+// 		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.(runtime.Unstructured).UnstructuredContent(), pod); err != nil {
+// 			utilruntime.HandleError(fmt.Errorf("error converting unstructured to pod: %w", err))
+// 		}
+//
+// 		return pod, true
+// 	case servicesGVR:
+// 		service := &corev1.Service{}
+// 		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.(runtime.Unstructured).UnstructuredContent(), service); err != nil {
+// 			utilruntime.HandleError(fmt.Errorf("error converting unstructured to service: %w", err))
+// 		}
+//
+// 		return service, true
+// 	case persistentVolumeClaimsGVR:
+// 		pvc := &corev1.PersistentVolumeClaim{}
+// 		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.(runtime.Unstructured).UnstructuredContent(), pvc); err != nil {
+// 			utilruntime.HandleError(fmt.Errorf("error converting unstructured to persistent volume claim: %w", err))
+// 		}
+//
+// 		return pvc, true
+// 	}
+//
+// 	return obj, false
+// }
+
+// getOrCreateDelegate creates a resourcequota.QuotaAdmission plugin for clusterName.
+func (k *KubeResourceQuota) getOrCreateDelegate(clusterName logicalcluster.Name) (*stoppableQuotaAdmission, error) {
+	var delegate *stoppableQuotaAdmission
+
+	k.lock.RLock()
+	delegate = k.delegates[clusterName]
+	k.lock.RUnlock()
+
+	if delegate != nil {
+		return delegate, nil
+	}
+
+	k.lock.Lock()
+	defer k.lock.Unlock()
+
+	delegate = k.delegates[clusterName]
+	if delegate != nil {
+		return delegate, nil
+	}
+
+	// Set up a context that is cancelable and that is bounded by k.ServerDone
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
Do we not have some higher context here for the server that's not `Background()`?

Reply:
That's actually what the goroutine right after this is doing. It's taking k.serverDone (which is actually the `ctx` passed to `Server.Run(ctx)` where `Server` is the main kcp server.... although ...

Looking at this again, I think I'm only passing this ctx's Done ch to resourcequota.NewResourceQuota, so I could probably scrap this and just do `resourcequota.NewResourceQuota(k.userSuppliedConfiguration, 5, k.serverDone)`, right?

Reply:
Oh wait, nevermind on that. I need to be able to call `cancel()` per instance when a ClusterWorkspace is deleted.

Reply:
But the goroutine as written probably isn't ideal - it lives as long as the server does, and ideally it terminates when a ClusterWorkspace is deleted too.